packages/@uppy/aws-s3/src/s3-client/S3.ts

```diff
@@ -27,37 +38,93 @@ class S3mini {
 
   constructor({
     endpoint,
     signRequest,
+    getCredentials,
     region = 'auto',
     requestSizeInBytes = C.DEFAULT_REQUEST_SIZE_IN_BYTES,
     requestAbortTimeout = undefined,
     fetch = globalThis.fetch,
   }: IT.S3Config) {
-    this._validateConstructorParams(endpoint, signRequest)
+    this._validateConstructorParams(endpoint, signRequest, getCredentials)
     this.endpoint = new URL(this._ensureValidUrl(endpoint))
-    this.signRequest = signRequest
     this.region = region
     this.requestSizeInBytes = requestSizeInBytes
     this.requestAbortTimeout = requestAbortTimeout
-    // Bind fetch to globalThis to preserve correct 'this' context in browsers
-    // Without this, calling this.fetch() throws "Illegal invocation"
     this.fetch = fetch.bind(globalThis)
+
+    if (signRequest) {
+      this.signRequest = signRequest
+    } else if (getCredentials) {
+      this.getCredentials = getCredentials
+      this.signRequest = this._createCredentialBasedSigner()
+    } else {
+      throw new TypeError(
+        'Either signRequest or getCredentials must be provided',
+      )
+    }
+  }
+
+  /** Creates a signer that fetches/caches credentials and signs requests. */
+  private _createCredentialBasedSigner(): IT.signRequestFn {
+    return async (request: IT.signableRequest): Promise<IT.signedHeaders> => {
+      const creds = await this._getCachedCredentials()
+      const signer = createSigV4Signer({
+        accessKeyId: creds.credentials.accessKeyId,
+        secretAccessKey: creds.credentials.secretAccessKey,
+        sessionToken: creds.credentials.sessionToken,
+        region: creds.region || this.region,
+      })
+      return signer(request)
+    }
+  }
+
+  /** Gets cached credentials or fetches new ones. Clears cache at 50% of expiry. */
+  private async _getCachedCredentials(): Promise<IT.CredentialsResponse> {
+    if (this.cachedCredentials == null) {
+      this.cachedCredentials = await this.getCredentials!()
+
+      // Clear cache at 50% of credential lifetime
+      const expiration = this.cachedCredentials.credentials.expiration
+      if (expiration) {
+        const expiresInMs = new Date(expiration).getTime() - Date.now()
+        if (expiresInMs > 0) {
+          setTimeout(() => {
+            this.cachedCredentials = undefined
+          }, expiresInMs / 2)
+        }
+      }
+    }
+    return this.cachedCredentials
   }
 
   private _validateConstructorParams(
     endpoint: string,
-    signRequest: IT.signRequestFn,
+    signRequest?: IT.signRequestFn,
+    getCredentials?: IT.getCredentialsFn,
   ): void {
     if (typeof endpoint !== 'string' || endpoint.trim().length === 0) {
       throw new TypeError(C.ERROR_ENDPOINT_REQUIRED)
     }
 
-    if (typeof signRequest !== 'function') {
-      throw new TypeError('signRequest is not passed')
+    if (!signRequest && !getCredentials) {
+      throw new TypeError(
+        'Either signRequest or getCredentials must be provided',
+      )
```

Comment:
This is duplicate, you already check the same thing in the constructor directly. I would probably remove the check in the constructor directly?

Reply:
yes the else block in the constructor where I'm throwing error is redundant will remove it.